inoculate/src/trace.rs

```diff
@@ -0,0 +1,168 @@
+    field::Visit,
+    fmt::{format::Writer, FmtContext, FormatEvent, FormatFields, FormattedFields},
+    registry::LookupSpan,
+};
+
+#[derive(Default)]
+pub struct CargoFormatter(());
+
+impl<S, N> FormatEvent<S, N> for CargoFormatter
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'a> FormatFields<'a> + 'static,
+{
+    fn format_event(
+        &self,
+        ctx: &FmtContext<'_, S, N>,
+        mut writer: Writer<'_>,
+        event: &Event<'_>,
+    ) -> fmt::Result {
+        let metadata = event.metadata();
+        let level = metadata.level();
+        let color = ColorMode::default().should_color_stdout();
+
+        let include_spans = {
+            let mut visitor = Visitor::new(*level, writer.by_ref(), color);
+            event.record(&mut visitor);
+            !visitor.did_cargo_format && ctx.lookup_current().is_some()
+        };
+
+        writer.write_char('\n')?;
+
+        if include_spans {
+            let pipe_style = if color {
+                style().blue().bold()
+            } else {
+                style()
+            };
+            let span_name = if color { style().bold() } else { style() };
+            writeln!(
+                writer,
+                "   {} {}",
+                "-->".style(pipe_style),
+                metadata.file().unwrap_or_else(|| metadata.target()),
+            )?;
+            ctx.visit_spans(|span| {
+                let exts = span.extensions();
+                let fields = exts
+                    .get::<FormattedFields<N>>()
+                    .map(|f| f.fields.as_str())
+                    .unwrap_or("");
+                writeln!(
+                    writer,
+                    "    {}  {}{}{}",
+                    "|".style(pipe_style),
+                    span.name().style(span_name),
+                    if fields.is_empty() { "" } else { ": " },
+                    fields
+                )
+            })?;
+
+            writer.write_char('\n')?;
+        }
+
+        Ok(())
+    }
+}
+
+struct Visitor<'writer> {
+    level: Level,
+    writer: Writer<'writer>,
+    is_empty: bool,
+    color: bool,
+    did_cargo_format: bool,
+}
+
+impl<'writer> Visitor<'writer> {
+    const MESSAGE: &'static str = "message";
+    const INDENT: usize = 12;
+
+    fn new(level: Level, writer: Writer<'writer>, color: bool) -> Self {
+        Self {
+            level,
+            writer,
+            is_empty: true,
+            did_cargo_format: false,
+            color,
+        }
+    }
+}
+
+impl Visit for Visitor<'_> {
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        if self.is_empty {
+            if self.level >= Level::INFO && field.name() == Self::MESSAGE {
+                let message = format!("{:?}", value);
+                if let Some((tag, message)) = message.as_str().split_once(' ') {
+                    if tag.len() <= Self::INDENT && tag.ends_with("ing") || tag.ends_with('d') {
```

Comment:
this code is definitely not sketchy at all :-)

I suppose doing it with separate fields for the action would be somewhat trickier as `MESSAGE` is usually the first field visited?

Reply:
idk, I kind of like this. alternatively, we could just drop the special casing of "ing" and just always emit the cargo-style formatting for INFO events whose first word is less than the max length. which might actually be nicer, idk.

Reply:
I kinda like the idea of requiring a pattern like `Action: message goes here`, where a `:` after the first word is stripped and used to determine that it was intended to be a cargo style info event, but I don't feel a strong opinion about it.